File: src/test/java/de/mwa/evopiaserver/ChannelAcceptanceTest.java
package de.mwa.evopiaserver;

import de.mwa.evopiaserver.api.dto.ChannelDto;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.json.AutoConfigureJson;
import org.springframework.boot.test.autoconfigure.json.AutoConfigureJsonTesters;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.util.TestPropertyValues;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.test.context.ContextConfiguration;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@Testcontainers
@AutoConfigureJson
@AutoConfigureJsonTesters
@ContextConfiguration(initializers = {ChannelAcceptanceTest.Initializer.class})
public class ChannelAcceptanceTest {
    @LocalServerPort
    private int port;

    @Autowired
    private TestRestTemplate restTemplate;

    @Autowired
    private RepositoryTestHelper repositoryTestHelper;

    @Container
    private static final PostgreSQLContainer postgreSQLContainer = new PostgreSQLContainer("postgres")
            .withDatabaseName("integration-tests-db")
            .withUsername("sa")
            .withPassword("sa");

    @AfterEach
    public void cleanup() {
        repositoryTestHelper.resetChannelTable();
    }

    @Test
    public void initial_channel_has_only_dummychannel() {
        List<ChannelDto> channels = getAllChannels();

        assertThat(channels).containsOnly(new ChannelDto("Dummychannel"));
    }

    @Test
    public void adding_removing_channel_should_work() {
        var addingUrl = "http://localhost:" + port + "/v2/channels/add";
        var body = "{\"name\": \"BatSign\"}";
        var addResponse = restTemplate.exchange
                (addingUrl, HttpMethod.POST, HttpEntityFactory.forTestUserWith(body), String.class);

        assertThat(addResponse.getStatusCode()).isEqualTo(HttpStatus.OK);

        var channels = getAllChannels();
        assertThat(channels).containsOnly(new ChannelDto("Dummychannel"), new ChannelDto("BatSign"));
    }

    @Test
    public void removing_channel_should_work() {
        var removingUrl = "http://localhost:" + port + "/v2/channels/remove";
        var body = "{\"name\": \"Dummychannel\"}";
        var removeResponse = restTemplate.exchange
                (removingUrl, HttpMethod.POST, HttpEntityFactory.forTestUserWith(body), String.class);

        assertThat(removeResponse.getStatusCode()).isEqualTo(HttpStatus.OK);

        var channelsAfterDelete = getAllChannels();
        assertThat(channelsAfterDelete).isEmpty();
    }

    private List<ChannelDto> getAllChannels() {
        var askingUrl = "http://localhost:" + port + "/v2/channels";
        var requestType = new ParameterizedTypeReference<List<ChannelDto>>() {};
        var askResponse = restTemplate.exchange
                (askingUrl, HttpMethod.GET, HttpEntityFactory.forTestUser(), requestType);

        assertThat(askResponse.getStatusCode()).isEqualTo(HttpStatus.OK);
        return askResponse.getBody();
    }

    static class Initializer
            implements ApplicationContextInitializer<ConfigurableApplicationContext> {
        public void initialize(ConfigurableApplicationContext configurableApplicationContext) {
            TestPropertyValues.of(
                    "spring.datasource.url=" + postgreSQLContainer.getJdbcUrl(),
                    "spring.datasource.username=" + postgreSQLContainer.getUsername(),
                    "spring.datasource.password=" + postgreSQLContainer.getPassword()
            ).applyTo(configurableApplicationContext.getEnvironment());
        }
    }
}

File: src/test/kotlin/de/mwa/evopiaserver/ChannelAcceptanceTest.kt
package de.mwa.evopiaserver

import de.mwa.evopiaserver.db.kotlin.de.mwa.evopiaserver.dto.ChannelDto
import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.http.HttpMethod.Companion.Delete
import io.ktor.http.HttpMethod.Companion.Post
import io.ktor.server.testing.*
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.assertThat
import org.junit.Assert.assertThat
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import org.testcontainers.containers.PostgreSQLContainer
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.junit.jupiter.Testcontainers


@OptIn(ExperimentalCoroutinesApi::class)
class ChannelAcceptanceTest : ServerTestSetup() {

    @AfterEach
    fun localCleanup() {
        repositoryTestHelper.resetChannelTable()
    }


    @Test
    fun initial_channel_has_only_dummychannel() = runTest {
        val channels = getAllChannels()

        assertThat(channels).containsOnly(ChannelDto("Dummychannel"))
    }

    @Test
    fun adding_channel_should_work() = runTest {
        val url = "http://localhost:8080/v3/channels/add"
        val body = "{\"name\": \"BatSign\"}"

        val addResponse = client.request(url) {
            method = Post
            setBody(body)
        }

        assertThat(addResponse.status).isEqualTo(HttpStatusCode.OK)

        val channels = getAllChannels()
        assertThat(channels).containsOnly(ChannelDto("Dummychannel"), ChannelDto("BatSign"))
    }

    @Test
    fun removing_channel_should_work() = runTest {
        val url = "http://localhost:8080/v3/channels/remove"
        val body = "{\"name\": \"Dummychannel\"}"

        val addResponse = client.request(url) {
            method = Post
            setBody(body)
        }

        assertThat(addResponse.status).isEqualTo(HttpStatusCode.OK)

        val channelsAfterDelete = getAllChannels()
        assertThat(channelsAfterDelete).isEmpty()
    }

    private suspend fun getAllChannels(): List<ChannelDto> {
        val url = "http://localhost:8080/v3/channels"

        val askResponse = client.request(url)

        assertThat(askResponse.status)
            .isEqualTo(HttpStatusCode.OK)

        return Json.decodeFromString(askResponse.bodyAsText())
    }
}
